File: src/cathedral/miner/submit.py
"""Miner-side claim submission."""

from __future__ import annotations

import os
from datetime import UTC, datetime

import httpx

from cathedral.config import MinerSettings
from cathedral.types import PolarisAgentClaim


class SubmitError(Exception):
    pass


async def submit_claim(
    settings: MinerSettings,
    *,
    work_unit: str,
    polaris_agent_id: str,
    polaris_deployment_id: str | None = None,
    polaris_run_ids: list[str] | None = None,
    polaris_artifact_ids: list[str] | None = None,
) -> int:
    bearer = os.environ.get(settings.validator_bearer_env)
    if not bearer:
        raise SubmitError(f"missing bearer env {settings.validator_bearer_env}")

    claim = PolarisAgentClaim(
        miner_hotkey=settings.miner_hotkey,
        owner_wallet=settings.owner_wallet,
        work_unit=work_unit,
        polaris_agent_id=polaris_agent_id,
        polaris_deployment_id=polaris_deployment_id,
        polaris_run_ids=polaris_run_ids or [],
        polaris_artifact_ids=polaris_artifact_ids or [],
        submitted_at=datetime.now(UTC),
    )

    url = f"{settings.validator_url.rstrip('/')}/v1/claim"
    async with httpx.AsyncClient(timeout=30.0) as client:
        try:
            r = await client.post(
                url,
                headers={"Authorization": f"Bearer {bearer}"},
                json=claim.model_dump(mode="json"),
            )
        except httpx.HTTPError as e:
            raise SubmitError(f"transport: {e}") from e

    if r.status_code >= 400:
        raise SubmitError(f"validator rejected claim: {r.status_code} {r.text}")
    body = r.json()
    return int(body.get("id", 0))

File: crates/cathedral-miner/src/submit.rs
use cathedral_types::{ClaimVersion, Hotkey, PolarisAgentClaim};

use crate::config::MinerConfig;

#[derive(Debug, thiserror::Error)]
pub enum SubmitError {
    #[error("transport: {0}")]
    Transport(String),
    #[error("validator rejected claim: status {status}, body: {body}")]
    Rejected { status: u16, body: String },
    #[error("missing bearer env {0}")]
    MissingBearer(String),
}

pub struct ClaimInputs<'a> {
    pub work_unit: &'a str,
    pub polaris_agent_id: &'a str,
    pub polaris_deployment_id: Option<&'a str>,
    pub polaris_run_ids: Vec<String>,
    pub polaris_artifact_ids: Vec<String>,
}

pub async fn submit_claim(cfg: &MinerConfig, inputs: ClaimInputs<'_>) -> Result<(), SubmitError> {
    let claim = PolarisAgentClaim {
        version: ClaimVersion::V1,
        miner_hotkey: Hotkey(cfg.miner_hotkey.clone()),
        owner_wallet: cfg.owner_wallet.clone(),
        work_unit: inputs.work_unit.to_string(),
        polaris_agent_id: inputs.polaris_agent_id.to_string(),
        polaris_deployment_id: inputs.polaris_deployment_id.map(str::to_string),
        polaris_run_ids: inputs.polaris_run_ids,
        polaris_artifact_ids: inputs.polaris_artifact_ids,
        submitted_at: chrono::Utc::now(),
    };

    let bearer = std::env::var(&cfg.validator_bearer_env)
        .map_err(|_| SubmitError::MissingBearer(cfg.validator_bearer_env.clone()))?;

    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(30))
        .build()
        .map_err(|e| SubmitError::Transport(e.to_string()))?;

    let url = format!("{}/v1/claim", cfg.validator_url.trim_end_matches('/'));
    let resp = client
        .post(url)
        .bearer_auth(bearer)
        .json(&claim)
        .send()
        .await
        .map_err(|e| SubmitError::Transport(e.to_string()))?;

    if resp.status().is_success() {
        Ok(())
    } else {
        let status = resp.status().as_u16();
        let body = resp.text().await.unwrap_or_default();
        Err(SubmitError::Rejected { status, body })
    }
}
